Serverless Monitoring Guide (#7532)
DataDog/documentation: content/en/infrastructure/serverless/serverless_monitoring_guide/nodejs.md

{{% /tab %}}
{{% /tabs %}}

## Graph Lambda Cold Starts

With both the Datadog Forwarder and Datadog Lambda Layer properly configured, we can navigate to the out-of-the-box [Enhanced Lambda Metrics dashboard][8].

From here, we immediately see a graph named “Cold Starts by Function”, which we can expand to get an in-depth overview of which functions we may, for example, want to provide additional provisioned concurrency to.

{{< img src="serverless/guides/end-to-end_serverless_monitoring_database/guide_video_2-compressed.mp4" video="true" alt="Enhanced Metrics Dashboard"  style="width:90%;">}}

You can easily recognize any enhanced Lambda metric as they are all named with the prefix `aws.lambda.enhanced.*`. For example, the above graph measures `aws.lambda.enhanced.invocations`, with the tag `cold_start:true`.

# Section 2: Navigate Lambda Error Logs

## Enable Lambda Log Ingestion

In the previous section, we set up the Datadog Forwarder to enable enhanced metrics. The Datadog Forwarder is also required to send logs from your Lambda function to Datadog.

{{< tabs >}}
{{% tab "Serverless Framework" %}}

> If you’ve been following along, the Datadog Forwarder is already enabled, and Lambda logs should already automatically be flowing through to Datadog. If you have been following step-by-step but logs aren’t appearing in Datadog, [refer to the Lambda logs troubleshooting steps][1].

Reviewer: ```suggestion
At this point, the Datadog Forwarder is enabled, and Lambda logs should already automatically be flowing through to Datadog. If  logs aren’t yet appearing in Datadog, [refer to the Lambda logs troubleshooting steps][1].
```

If something is already subscribing to a Log Group that you want to monitor with the Datadog Lambda Function, you can add the ARN of the Forwarder to the ‘forwarder’ field in your serverless.yml file. When set, the plugin will try to subscribe the function’s CloudWatch log groups to the Forwarder. For example, your serverless.yml could now look like this:

```bash
custom:
    datadog:
        addLayers: true
        flushMetricsToLogs: true
        logLevel: 'INFO'
        enableDDTracing: true
        # Add the ARN of your Forwarder.
        forwarder:
```

[1]: https://docs.datadoghq.com/logs/guide/lambda-logs-collection-troubleshooting-guide/

{{% /tab %}}
{{% tab "AWS Console" %}}

> If you’ve been following along, the Datadog Forwarder is already enabled, and Lambda logs should already automatically be flowing through to Datadog. If you have been following step-by-step but logs aren’t appearing in Datadog, [refer to the Lambda logs troubleshooting steps][1].

Reviewer: ```suggestion
At this point, the Datadog Forwarder is enabled, and Lambda logs should already automatically be flowing through to Datadog. If  logs aren’t yet appearing in Datadog, [refer to the Lambda logs troubleshooting steps][1].
```

If something is already subscribing to a Log Group that you want to monitor with the Datadog Lambda Function, you can remove it from the AWS console:  

1. Select the log source on your Lambda function.
2. Select Remove Subscription Filter in the Actions pulldown

[1]: https://docs.datadoghq.com/logs/guide/lambda-logs-collection-troubleshooting-guide/

{{% /tab %}}
{{% /tabs %}}

If your logs still aren’t appearing in Datadog, you can [follow the troubleshooting steps here][9].  

## Triage your Lambda Errors in the Log Explorer

> We can visit the Datadog Serverless view again to view all of our Lambda functions. Now, if we navigate to a specific function, we are able to see the logs being emitted from the function. From here, we can navigate to the Log Explorer to learn more about the log.

Reviewer: ```suggestion
Visit the Datadog Serverless view to view all of the Lambda functions. Navigate to a specific function to see the logs being emitted from the function. Then navigate to the Log Explorer to learn more about the log.
```

{{< img src="serverless/guides/end-to-end_serverless_monitoring_database/guide_video_3-compressed.mp4" video="true" alt="Serverless Page Logs"  style="width:90%;">}}

You may be interested in knowing which errors are most frequently impacting your users. You can quickly filter through your logs to find the patterns of error logs surfacing from your Lambda functions in order of volume.

{{< img src="serverless/guides/end-to-end_serverless_monitoring_database/guide_video_4-compressed.mp4" video="true" alt="Log Patterns"  style="width:90%;">}}

# Section 3: Root Cause Analysis for Lambda Functions

## Enable Lambda Distributed Tracing

In the first section, we set up the Datadog Forwarder and Lambda Layer to enable enhanced metrics. Both the Datadog Forwarder and Lambda Layer are also required to surface distributed traces from your functions in Datadog.

{{< tabs >}}
{{% tab "Serverless Framework" %}}

> If you’ve been following along, the Datadog Forwarder and Lambda Layer are already enabled. With the Serverless Framework, traces are automatically flowing through to Datadog without any additional code instrumentation.

Reviewer: ```suggestion
At this point, the Datadog Forwarder is and Lambda Layer are already enabled. With the Serverless Framework, traces are automatically flowing through to Datadog without any additional code instrumentation.
```

{{% /tab %}}
{{% tab "Code Instrumentation" %}}

> If you’ve been following along, the Datadog Forwarder is enabled and the Lambda Layer has been installed on your function. Follow these steps to set up APM in Node:  

Reviewer: ```suggestion
Follow these steps to set up APM in Node:  
```

1. Manually add the tracing library to your project:

```bash
npm install datadog-lambda-js
npm install dd-trace

yarn add datadog-lambda-js
yarn add dd-trace
```

2. Instrument your code.

```js
const { datadog } = require('datadog-lambda-js');
const tracer = require('dd-trace').init(); // Any manual tracer config goes here.

// This function will be wrapped in a span
const longCalculation = tracer.wrap('calculation-long-number', () => {
        // An expensive calculation goes here
});

// This function will also be wrapped in a span
module.exports.hello = datadog((event, context, callback) => {
        longCalculation();

        callback(null, {
            statusCode: 200,
            body: 'Hello from serverless!'
        });
});
```

To instrument your Node libraries and customize your traces, consult the [Datadog Node APM documentation][1].

[1]: https://docs.datadoghq.com/tracing/setup/nodejs/

{{% /tab %}}
{{% /tabs %}}

## Root Cause Analysis using Traces, Metrics and Logs

> We can visit the Datadog Serverless page one more time to view all of our Lambda functions. If we navigate to a specific function, we are able to see the traces being emitted from the function. We can expand each trace to view a flame graph of the duration of the total request across Lambda functions, as well as correlated logs and Lambda metrics at the time of the request.

Reviewer: ```suggestion
View the Datadog Serverless page to see the Lambda functions. Navigate to a specific function to see the traces being emitted from the function. Expand each trace to view a flame graph of the duration of the total request across Lambda functions, and correlated logs and Lambda metrics at the time of the request.
```

{{< img src="serverless/guides/end-to-end_serverless_monitoring_database/guide_video_5-compressed.mp4" video="true" alt="Flame Graph"  style="width:90%;">}}

> We can use the traces surfaced in the flame graph to identify that the current Lambda function has an error due to an error that surfaced from a subsequent function call. We can quickly navigate to that function in the request, and find the error trace and the corresponding error log, which we identify to be a timeout. 

Reviewer: ```suggestion
Use the traces surfaced in the flame graph to identify that the current Lambda function has an error due to an error that surfaced from a subsequent function call. Navigate to that function in the request, and find the error trace and the corresponding error log, which in this example is a timeout. 
```

{{< img src="serverless/guides/end-to-end_serverless_monitoring_database/guide_video_6-compressed.mp4" video="true" alt="Root Cause Analysis"  style="width:90%;">}}

> With logs, traces and enhanced metrics all in one place, we are able to efficiently identify the root cause of an issue across our serverless requests.

Reviewer: ```suggestion
With logs, traces, and enhanced metrics all in one place, we are able to efficiently identify the root cause of an issue across our serverless requests.
```

For example, Datadog has a [DynamoDB integration][11] which collects DynamoDB table performance metrics and logs. Datadog also offers an Amazon SQS integration to ingest SQS metrics and to collect logs during SQS API calls. Both integrations are paired respectively with out-of-the-box dashboards for [DynamoDB tables][11] and [SQS][12].  Integrations for other common serverless services are listed below:  
  
**Data Stores:** RDS, Aurora, self-hosted, S3  
**Message Queues:** SNS, SQS, Kinesis  
**Lambda Features:** [AWS Step Functions][13]  

You can customize any of the default dashboards offered by the above integrations by pulling information from different services all into the same dashboard.

## Submit Custom Metrics

Custom metrics give additional insights into use cases that are unique to your application workflows, such as a user logging into your application, purchasing an item, or updating a user profile.

In the first section, we set up the Datadog Lambda Layer to enable enhanced metrics. The Datadog Lambda Layer is required to surface distributed traces from your functions in Datadog.

> If you’ve been following along, the Datadog Lambda Layer is already enabled. Custom metrics sent from the Datadog Lambda Layer are automatically created as distributions, so you can graph the avg, sum, max, min, and count out-of-the-box. You can also enable your own percentiles in the Distribution Metrics page.

Reviewer: ```suggestion
Once you have the Datadog Lambda Layer enabled, custom metrics are automatically created as distributions, so you can graph the `avg`, `sum`, `max`, `min`, and `count` out-of-the-box. You can also enable your own percentiles in the Distribution Metrics page.
```

```

{{% /tab %}}
{{% tab "AWS Console" %}}

To ingest a custom metric from your function:  

1. Ensure you have the AWS Lambda environment variable `DD_FLUSH_TO_LOG` set to `true`.  
2. In your function code, you must import the necessary methods from the Lambda Layer and add a wrapper around your function handler:

```js
const { datadog, sendDistributionMetric } = require('datadog-lambda-js');

async function myHandler(event, context) {
        sendDistributionMetric(
            'coffee_house.order_value', // Metric name
            12.45, // Metric value
            'product:latte',
            'order:online' // Associated tags
        );
        return {
            statusCode: 200,
            body: 'hello, dog!'
        };
}
// You only need to wrap your function handler (Not helper functions).
module.exports.myHandler = datadog(myHandler);
```

{{% /tab %}}
{{% /tabs %}}

Creating a monitor for a metric enables you to get notified about key issues in your serverless applications. You could then determine root cause as illustrated earlier in this guide by combining metrics, traces, and logs across your Lambda functions, data stores, and message queues.

> {{< img src="serverless/guides/end-to-end_serverless_monitoring_database/guide_video_7-compressed.mp4" video="true" alt="Custom Metrics"  style="width:90%;">}}

Reviewer 2: Similar feedback, let's reshoot this so we can see what's going on

> Creating a monitor for a metric enables you to get notified about key issues in your serverless applications. You could then determine root cause as illustrated earlier in this guide by combining metrics, traces and logs across your Lambda functions, data stores, and message queues.

Reviewer: ```suggestion
Creating a monitor for a metric enables you to get notified about key issues in your serverless applications. You could then determine root cause as illustrated earlier in this guide by combining metrics, traces, and logs across your Lambda functions, data stores, and message queues.
```
